### PR TITLE
Test CUDA in all supported Standard modes.

### LINKED ISSUE #4965: Test CUDA in all supported Standard modes
Currently, we test CUDA in only C++14 mode, because that's all that was available when we added this test coverage years ago:

https://github.com/microsoft/STL/blob/a26f4adb1ecf767784bb4e4318093c4579fdb364/tests/std/tests/GH_000639_nvcc_include_all/env.lst#L4-L7

Since then, they've added support for C++17 and C++20. We should extend this test accordingly (after figuring out the right compiler options).

### PR DESCRIPTION
**Pull request by fsb4000**
Fixes #4965 

1) I didn't figure out why `-std=c++17` is failing without `D_SILENCE_ALL_CXX17_DEPRECATION_WARNINGS` and `-std=c++20` is not failing.
I got the error:
```
C:\Dev\STL\out\x64\out\inc\ccomplex(21): error C2220: the following warning is treated as an error
C:\Dev\STL\out\x64\out\inc\ccomplex(21): warning C4996: '_Header_ccomplex': warning STL4004: <ccomplex>, <cstdalign>, <cstdbool>, and <ctgmath> are deprecated in C++17. You can define _SILENCE_CXX17_C_HEADER_DEPRECATION_WARNING or _SILENCE_ALL_CXX17_DEPRECATION_WARNINGS to suppress this warning.
C:\Dev\STL\out\x64\out\inc\cstdalign(22): warning C4996: '_Header_cstdalign': warning STL4004: <ccomplex>, <cstdalign>, <cstdbool>, and <ctgmath> are deprecated in C++17. You can define _SILENCE_CXX17_C_HEADER_DEPRECATION_WARNING or _SILENCE_ALL_CXX17_DEPRECATION_WARNINGS to suppress this warning.
C:\Dev\STL\out\x64\out\inc\cstdbool(21): warning C4996: '_Header_cstdbool': warning STL4004: <ccomplex>, <cstdalign>, <cstdbool>, and <ctgmath> are deprecated in C++17. You can define _SILENCE_CXX17_C_HEADER_DEPRECATION_WARNING or _SILENCE_ALL_CXX17_DEPRECATION_WARNINGS to suppress this warning.
C:\Dev\STL\out\x64\out\inc\ctgmath(22): warning C4996: '_Header_ctgmath': warning STL4004: <ccomplex>, <cstdalign>, <cstdbool>, and <ctgmath> are deprecated in C++17. You can define _SILENCE_CXX17_C_HEADER_DEPRECATION_WARNING or _SILENCE_ALL_CXX17_DEPRECATION_WARNINGS to suppress this warning.
```

2) C++23 is not supported as far as I tried:
```
$ nvcc -std=c++23
nvcc fatal   : Value 'c++23' is not defined for option 'std'
```
```
$ nvcc -std=c++2b
nvcc fatal   : Value 'c++2b' is not defined for option 'std'
```

3) I found that we ignore all nvcc warnings.  If I add `--diag-error=128` I get:
```
C:/Dev/STL/out/x64/out/inc\xutility(5069): error: loop is not reachable
          for (; _First != _Last; ++_Dest, (void) ++_First) {
          ^
          detected during:
            instantiation of "_OutIt std::_Move_unchecked(_InIt, _InIt, _OutIt) [with _InIt=std::seed_seq::result_type *, _OutIt=std::seed_seq::result_type *]" at line 1756 of C:/Dev/STL/out/x64/out/inc\vector
            instantiation of "std::vector<_Ty, _Alloc>::iterator std::vector<_Ty, _Alloc>::erase(std::vector<_Ty, _Alloc>::const_iterator, std::vector<_Ty, _Alloc>::const_iterator) [with _Ty=std::seed_seq::result_type, _Alloc=std::allocator<std::seed_seq::result_type>]" at line 3541 of C:/Dev/STL/out/x64/out/inc\vector
            instantiation of "void std::vector<__nv_bool, _Alloc>::_Trim(std::vector<__nv_bool, _Alloc>::size_type) [with _Alloc=std::allocator<__nv_bool>]" at line 2886 of C:/Dev/STL/out/x64/out/inc\vector
            instantiation of "std::vector<__nv_bool, _Alloc>::vector(std::vector<__nv_bool, _Alloc>::size_type, const _Alloc &) [with _Alloc=std::allocator<__nv_bool>]" at line 681 of C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.41.34120\include\ppltasks.h

C:/Dev/STL/out/x64/out/inc\xutility(4735): error: loop is not reachable
      for (; _Count != 0; ++_Dest, (void) ++_First, --_Count) {
      ^
          detected during:
            instantiation of "_OutIt std::_Copy_n_unchecked4(_InIt, _SizeTy, _OutIt) [with _InIt=void **, _SizeTy=size_t, _OutIt=void **]" at line 1430 of C:/Dev/STL/out/x64/out/inc\vector
            instantiation of "void std::vector<_Ty, _Alloc>::_Assign_counted_range(_Iter, std::vector<_Ty, _Alloc>::size_type) [with _Ty=void *, _Alloc=std::allocator<void *>, _Iter=void **]" at line 1492 of C:/Dev/STL/out/x64/out/inc\vector
            instantiation of "std::vector<_Ty, _Alloc> &std::vector<_Ty, _Alloc>::operator=(const std::vector<_Ty, _Alloc> &) [with _Ty=void *, _Alloc=std::allocator<void *>]" at line 181 of C:\Program Files\Microsoft Visual Studio\2022\Preview\VC\Tools\MSVC\14.41.34120\include\ppltasks.h

10 errors detected in the compilation of "C:/Dev/STL/tests/std/tests/GH_000639_nvcc_include_all/test.compile.pass.cpp".
```
 `--diag-error=177`
```
C:\Dev\STL\out\x64\out\inc\memory_resource(586): error: variable "_Guard" was declared but never referenced
              lock_guard<mutex> _Guard{_Mtx};
                                ^

C:\Dev\STL\out\x64\out\inc\condition_variable(66): error: variable "_Guard" was declared but never referenced
          lock_guard<mutex> _Guard{*_Myptr};
                            ^

C:\Dev\STL\out\x64\out\inc\condition_variable(71): error: variable "_Guard" was declared but never referenced
          lock_guard<mutex> _Guard{*_Myptr};
                            ^

C:\Dev\STL\out\x64\out\inc\shared_mutex(92): error: variable "_Lock" was declared but never referenced
          lock_guard<mutex> _Lock(_Mymtx);
                            ^

C:\Dev\STL\out\x64\out\inc\shared_mutex(133): error: variable "_Lock" was declared but never referenced
              lock_guard<mutex> _Lock(_Mymtx);
                                ^

C:\Dev\STL\out\x64\out\inc\shared_mutex(151): error: variable "_Lock" was declared but never referenced
          lock_guard<mutex> _Lock(_Mymtx);
                            ^

C:\Dev\STL\out\x64\out\inc\shared_mutex(187): error: variable "_Lock" was declared but never referenced
              lock_guard<mutex> _Lock(_Mymtx);
                                ^

45 errors detected in the compilation of "C:/Dev/STL/tests/std/tests/GH_000639_nvcc_include_all/test.compile.pass.cpp".
```
 `--diag-error=186`
```
C:\Dev\STL\out\x64\out\inc\xstring(2401): error: pointless comparison of unsigned integer with zero
          do { if (_Result_size >= 0) { } else { do { (void) ((1 != _CrtDbgReport(2, "C:\\Dev\\STL\\out\\x64\\out\\inc\\xstring", 2401, 0, "%s", "the returned size can't be smaller than 0")) || (__debugbreak(), 0)); ::_invalid_parameter(L"\"the returned size can't be smaller than 0\"", L"", L"C:\\Dev\\STL\\out\\x64\\out\\inc\\xstring", static_cast<unsigned int>(2401), 0); } while (false); } ; } while (false);
                                ^
          detected during instantiation of "void std::basic_string<_Elem, _Traits, _Alloc>::_Resize_and_overwrite(std::basic_string<_Elem, _Traits, _Alloc>::size_type, _Operation) [with _Elem=__wchar_t, _Traits=std::char_traits<__wchar_t>, _Alloc=std::allocator<__wchar_t>, _Operation=lambda [](__wchar_t *, size_t)->size_t]" at line 1405 of C:\Dev\STL\out\x64\out\inc\filesystem

1 error detected in the compilation of "C:/Dev/STL/tests/std/tests/GH_000639_nvcc_include_all/test.compile.pass.cpp".
test.compile.pass.cpp
```
I didn't find how to enable all/almost all warnings with one command and if the warnings are good. We can enable three with `--diag-error=128,177,186` but I don't want to add string like this`--diag-error=1,2,3,4,5,...,200`. And  I found this:
https://github.com/microsoft/STL/blob/1e312b38db8df1dfbea17adc344454feb8d00dd9/stl/inc/string#L459-L469
So we took care about some nvcc warnings in the past...

4) About the `typename`: https://gcc.godbolt.org/z/6TcboYYEe so I think the issue is fixed in CUDA 12.5.1, so we don't need to report upstream to Nvidia.

### CODE REVIEW

File: tests/std/tests/GH_000639_nvcc_include_all/env.lst [hunk 1]
```diff
@@ -5,3 +5,8 @@ PM_COMPILER="nvcc" PM_CL="--x cu -Xcompiler -Od,-EHsc,-nologo,-W4,-WX,-openmp"
 RUNALL_CROSSLIST
 PM_CL="-Xcompiler -MT"
 PM_CL="--debug -Xcompiler -MTd"
+RUNALL_CROSSLIST
+PM_CL=""
+PM_CL="-std=c++14"
+PM_CL="-std=c++17 -D_SILENCE_ALL_CXX17_DEPRECATION_WARNINGS"
```
Comment (AlexGuteniev) [inline tests/std/tests/GH_000639_nvcc_include_all/env.lst:11]: Isn't `_SILENCE_CXX17_C_HEADER_DEPRECATION_WARNING` satisfactory?

File: tests/std/tests/GH_000639_nvcc_include_all/env.lst [hunk 2]
```diff
@@ -5,3 +5,8 @@ PM_COMPILER="nvcc" PM_CL="--x cu -Xcompiler -Od,-EHsc,-nologo,-W4,-WX,-openmp"
 RUNALL_CROSSLIST
 PM_CL="-Xcompiler -MT"
 PM_CL="--debug -Xcompiler -MTd"
+RUNALL_CROSSLIST
+PM_CL=""
```
Comment (AlexGuteniev) [inline tests/std/tests/GH_000639_nvcc_include_all/env.lst:9]: Is no language mode different from each of the language modes?

Comment (fsb4000) [reply tests/std/tests/GH_000639_nvcc_include_all/env.lst:9]: no language mode is the same as `-std=c++14`.

File: tests/std/tests/GH_000639_nvcc_include_all/env.lst (hunk 1 above)
Comment (fsb4000) [reply tests/std/tests/GH_000639_nvcc_include_all/env.lst:11]: With `-D_SILENCE_CXX17_C_HEADER_DEPRECATION_WARNING ` I got:
```
C:\Dev\STL\out\x64\out\inc\__msvc_all_public_headers.hpp(27): error C2220: the following warning is treated as an error
C:\Dev\STL\out\x64\out\inc\__msvc_all_public_headers.hpp(27): warning C4005: '_SILENCE_CXX17_C_HEADER_DEPRECATION_WARNING': macro redefinition
C:\Dev\STL\out\x64\out\inc\__msvc_all_public_headers.hpp(27): note: '_SILENCE_CXX17_C_HEADER_DEPRECATION_WARNING' previously declared on the command line
```
`__msvc_all_public_headers.hpp` has:
https://github.com/microsoft/STL/blob/1e312b38db8df1dfbea17adc344454feb8d00dd9/stl/inc/__msvc_all_public_headers.hpp#L27
So I don't understand why we have the errors without `D_SILENCE_ALL_CXX17_DEPRECATION_WARNINGS` at all.

Comment (frederick-vs-ja) [reply tests/std/tests/GH_000639_nvcc_include_all/env.lst:11]: Hmm... should we use the `#ifndef [...] #define [...] #endif` pattern in `__msvc_all_public_headers.hpp`?

Comment (StephanTLavavej) [reply tests/std/tests/GH_000639_nvcc_include_all/env.lst:11]: Is something force-including the headers before `__msvc_all_public_headers.hpp` gets a chance to define its macros?

File: tests/std/tests/GH_000639_nvcc_include_all/env.lst (hunk 2 above)
Comment (StephanTLavavej) [reply tests/std/tests/GH_000639_nvcc_include_all/env.lst:9]: If we're explicitly testing `-std=c++14` on the line below, then we shouldn't have an empty set here - that's redundant. (Elsewhere, our configs either use the pattern "nothing vs. something" or "explicitly specify all modes", and we use the latter for Standard modes.)

File: tests/std/tests/GH_000639_nvcc_include_all/env.lst (hunk 1 above)
Comment (fsb4000) [reply tests/std/tests/GH_000639_nvcc_include_all/env.lst:10]: Yes, `nvcc` force includes `cuda_runtime.h`.

I ran with `--verbose` and I got:
```
#$ resource file C:\Temp/tmpxft_00002f0c_00000000-11.res: [-D__CUDA_ARCH_LIST__=520 -D__NV_LEGACY_LAUNCH -nologo -E -TP -EHsc -D__CUDACC__ -D__NVCC__  -Od -EHsc -nologo -W4 -WX -openmp -MTd -I"C:/Dev/STL/out/x64/out/inc" -I"C:/Dev/STL/llvm-project/libcxx/test/support" -I"C:/Dev/STL/tests/std/include" "-IC:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.4\bin/../include"    -D__CUDACC_VER_MAJOR__=12 -D__CUDACC_VER_MINOR__=4 -D__CUDACC_VER_BUILD__=99 -D__CUDA_API_VER_MAJOR__=12 -D__CUDA_API_VER_MINOR__=4 -D__NVCC_DIAG_PRAGMA_SUPPORT__=1 -FI "cuda_runtime.h" -Zi "C:\Dev\STL\tests\std\tests\GH_000639_nvcc_include_all\test.compile.pass.cpp" ]
#$ cl.exe /std:c++17 @"C:\Temp/tmpxft_00002f0c_00000000-11.res" > "C:/Temp/tmpxft_00002f0c_00000000-9_test.compile.pass.cpp4.ii"
test.compile.pass.cpp
#$ erase C:\Temp/tmpxft_00002f0c_00000000-11.res
#$ cudafe++ --ms_c++17 --microsoft_version=1941 --msvc_target_version=1941 --compiler_bindir "C:/Program Files/Microsoft Visual Studio/2022/Preview/VC/Tools/MSVC/14.41.34120/bin/HostX64/x64/../../../../../../.." --sdk_dir "C:/Program Files (x86)/Windows Kits/10/" --diag_suppress 128,177,186 --display_error_number --orig_src_file_name "C:/Dev/STL/tests/std/tests/GH_000639_nvcc_include_all/test.compile.pass.cpp" --orig_src_path_name "C:\Dev\STL\tests\std\tests\GH_000639_nvcc_include_all\test.compile.pass.cpp" --allow_managed --m64 --parse_templates --gen_c_file_name "C:/Temp/tmpxft_00002f0c_00000000-10_test.compile.pass.cudafe1.cpp" --stub_file_name "tmpxft_00002f0c_00000000-10_test.compile.pass.cudafe1.stub.c" --gen_module_id_file --module_id_file_name "C:/Temp/tmpxft_00002f0c_00000000-8_test.compile.pass.module_id" "C:/Temp/tmpxft_00002f0c_00000000-9_test.compile.pass.cpp4.ii"
#$ resource file C:\Temp/tmpxft_00002f0c_00000000-13.res: [-D__CUDA_ARCH__=520 -D__CUDA_ARCH_LIST__=520 -D__NV_LEGACY_LAUNCH -nologo -E -TP  -DCUDA_DOUBLE_MATH_FUNCTIONS -EHsc -D__CUDACC__ -D__NVCC__  -Od -EHsc -nologo -W4 -WX -openmp -MTd -I"C:/Dev/STL/out/x64/out/inc" -I"C:/Dev/STL/llvm-project/libcxx/test/support" -I"C:/Dev/STL/tests/std/include" "-IC:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.4\bin/../include"    -D__CUDACC_VER_MAJOR__=12 -D__CUDACC_VER_MINOR__=4 -D__CUDACC_VER_BUILD__=99 -D__CUDA_API_VER_MAJOR__=12 -D__CUDA_API_VER_MINOR__=4 -D__NVCC_DIAG_PRAGMA_SUPPORT__=1 -FI "cuda_runtime.h" -Zi "C:\Dev\STL\tests\std\tests\GH_000639_nvcc_include_all\test.compile.pass.cpp" ]
#$ cl.exe /std:c++17 @"C:\Temp/tmpxft_00002f0c_00000000-13.res" > "C:/Temp/tmpxft_00002f0c_00000000-12_test.compile.pass.cpp1.ii"
test.compile.pass.cpp
#$ erase C:\Temp/tmpxft_00002f0c_00000000-13.res
#$ cicc --ms_c++17 --microsoft_version=1941 --msvc_target_version=1941 --compiler_bindir "C:/Program Files/Microsoft Visual Studio/2022/Preview/VC/Tools/MSVC/14.41.34120/bin/HostX64/x64/../../../../../../.." --sdk_dir "C:/Program Files (x86)/Windows Kits/10/" --diag_suppress 128,177,186 --display_error_number --orig_src_file_name "C:/Dev/STL/tests/std/tests/GH_000639_nvcc_include_all/test.compile.pass.cpp" --orig_src_path_name "C:\Dev\STL\tests\std\tests\GH_000639_nvcc_include_all\test.compile.pass.cpp" --allow_managed  -arch compute_52 -m64 --no-version-ident -ftz=0 -prec_div=1 -prec_sqrt=1 -fmad=1 --include_file_name "tmpxft_00002f0c_00000000-7_test.compile.pass.fatbin.c" -tused --module_id_file_name "C:/Temp/tmpxft_00002f0c_00000000-8_test.compile.pass.module_id" --gen_c_file_name "C:/Temp/tmpxft_00002f0c_00000000-10_test.compile.pass.cudafe1.c" --stub_file_name "C:/Temp/tmpxft_00002f0c_00000000-10_test.compile.pass.cudafe1.stub.c" --gen_device_file_name "C:/Temp/tmpxft_00002f0c_00000000-10_test.compile.pass.cudafe1.gpu"  "C:/Temp/tmpxft_00002f0c_00000000-12_test.compile.pass.cpp1.ii" -o "C:/Temp/tmpxft_00002f0c_00000000-10_test.compile.pass.ptx"
#$ ptxas -arch=sm_52 -m64 "C:/Temp/tmpxft_00002f0c_00000000-10_test.compile.pass.ptx"  -o "C:/Temp/tmpxft_00002f0c_00000000-14_test.compile.pass.sm_52.cubin"
#$ fatbinary -64 --cicc-cmdline="-ftz=0 -prec_div=1 -prec_sqrt=1 -fmad=1 " "--image3=kind=elf,sm=52,file=C:/Temp/tmpxft_00002f0c_00000000-14_test.compile.pass.sm_52.cubin" "--image3=kind=ptx,sm=52,file=C:/Temp/tmpxft_00002f0c_00000000-10_test.compile.pass.ptx" --embedded-fatbin="C:/Temp/tmpxft_00002f0c_00000000-7_test.compile.pass.fatbin.c"
#$ erase C:/Temp/tmpxft_00002f0c_00000000-7_test.compile.pass.fatbin
#$ resource file C:\Temp/tmpxft_00002f0c_00000000-15.res: [-D__CUDA_ARCH__=520 -D__CUDA_ARCH_LIST__=520 -D__NV_LEGACY_LAUNCH -nologo -c -TP  -DCUDA_DOUBLE_MATH_FUNCTIONS -EHsc -Od -EHsc -nologo -W4 -WX -openmp -MTd -I"C:/Dev/STL/out/x64/out/inc" -I"C:/Dev/STL/llvm-project/libcxx/test/support" -I"C:/Dev/STL/tests/std/include" "-IC:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.4\bin/../include"   -Zi "C:\Temp\tmpxft_00002f0c_00000000-10_test.compile.pass.cudafe1.cpp" ]
#$ cl.exe /std:c++17 @"C:\Temp/tmpxft_00002f0c_00000000-15.res" -Fo"test.compile.pass.obj"
tmpxft_00002f0c_00000000-10_test.compile.pass.cudafe1.cpp
C:\Dev\STL\out\x64\out\inc\ccomplex(21): error C2220: the following warning is treated as an error
C:\Dev\STL\out\x64\out\inc\ccomplex(21): warning C4996: '_Header_ccomplex': warning STL4004: <ccomplex>, <cstdalign>, <cstdbool>, and <ctgmath> are deprecated in C++17. You can define _SILENCE_CXX17_C_HEADER_DEPRECATION_WARNING or _SILENCE_ALL_CXX17_DEPRECATION_WARNINGS to suppress this warning.
C:\Dev\STL\out\x64\out\inc\cstdalign(22): warning C4996: '_Header_cstdalign': warning STL4004: <ccomplex>, <cstdalign>, <cstdbool>, and <ctgmath> are deprecated in C++17. You can define _SILENCE_CXX17_C_HEADER_DEPRECATION_WARNING or _SILENCE_ALL_CXX17_DEPRECATION_WARNINGS to suppress this warning.
C:\Dev\STL\out\x64\out\inc\cstdbool(21): warning C4996: '_Header_cstdbool': warning STL4004: <ccomplex>, <cstdalign>, <cstdbool>, and <ctgmath> are deprecated in C++17. You can define _SILENCE_CXX17_C_HEADER_DEPRECATION_WARNING or _SILENCE_ALL_CXX17_DEPRECATION_WARNINGS to suppress this warning.
C:\Dev\STL\out\x64\out\inc\ctgmath(22): warning C4996: '_Header_ctgmath': warning STL4004: <ccomplex>, <cstdalign>, <cstdbool>, and <ctgmath> are deprecated in C++17. You can define _SILENCE_CXX17_C_HEADER_DEPRECATION_WARNING or _SILENCE_ALL_CXX17_DEPRECATION_WARNINGS to suppress this warning.
# --error 0x2 --
```
look at  `-FI "cuda_runtime.h"`

I looked at the file. And it has size of 91kb. and it includes some STL files.
Look at `#include <utility>`

изображение

**Review by CaseyCarter — APPROVED** [review]

**Comment by CaseyCarter** [comment]
I pushed a couple of commits to (1) address a comment nitpick, and (2) resolve the deprecation macro question by guarding the definition in `__msvc_all_public_headers.hpp` so we can define the more precise macro on the command line.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks! I'm still mystified as to why C++20 mode isn't emitting the C++17 deprecation warnings for C headers, or C++20's own deprecation warning for `<ciso646>`, but C++20 mode is clearly activated (since otherwise the `typename` workaround in `concept _Cpp17_random_delta` would not have been necessary), so I don't think this PR needs to be blocked over the mysterious absence of warnings.

**Comment by CaseyCarter** [comment]
> Thanks! I'm still mystified as to why C++20 mode isn't emitting the C++17 deprecation warnings for C headers, or C++20's own deprecation warning for `<ciso646>`, but C++20 mode is clearly activated (since otherwise the `typename` workaround in `concept _Cpp17_random_delta` would not have been necessary), so I don't think this PR needs to be blocked over the mysterious absence of warnings.

Wild guess: some predefined global warning suppression to avoid the force-included header emitting warnings?

**Comment by fsb4000** [comment]
> Thanks! I'm still mystified as to why C++20 mode isn't emitting the C++17 deprecation warnings for C headers

My guess: these headers are `deprecated in C++17` but `removed in C++20`, so probably `cuda_runtime.h` do something different for C++20 and this is why we don't see the warnings.

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for this major improvement in test coverage! :white_check_mark: :chart_with_upwards_trend: :heart_eyes_cat:

**Pull request merged**